internal/iapl/rbac.go

```diff
@@ -0,0 +1,142 @@
+package iapl
+
+import (
+	"go.infratographer.com/permissions-api/internal/types"
+)
+
+const (
+	// RoleOwnerRelation is the name of the relationship that connects a role to its owner.
+	RoleOwnerRelation = "owner"
+	// RoleOwnerMemberRoleRelation is the name of the relationship that connects a resource
+	// to a role that it owns
+	RoleOwnerMemberRoleRelation = "member_role"
+	// AvailableRoleRelation is the name of the relationship that list all roles
+	// that are available for a given resource
+	AvailableRoleRelation = "avail_role"
+	// RolebindingRoleRelation is the name of the relationship that connects a role binding to a role.
+	RolebindingRoleRelation = "role"
+	// RolebindingSubjectRelation is the name of the relationship that connects a role binding to a subject.
+	RolebindingSubjectRelation = "subject"
+	// RoleOwnerParentRelation is the name of the relationship that connects a role's owner to its parent.
+	RoleOwnerParentRelation = "parent"
+)
+
+// ResourceRoleBindingV2 describes the relationships that will be created
+// for a resource to support role-binding V2
+type ResourceRoleBindingV2 struct {
+	// AvailableRolesFrom is the list of resource types that can provide roles to this resource
+	AvailableRolesFrom []string
+}
+
+/*
+RBAC represents a role-based access control policy.
+  - RoleResource is the name of the resource type that represents a role.
+  - RoleRelationshipSubject is the name of the relationship that connects a role to a subject.
+  - RoleOwners is the names of the resource types that can own a role.
+  - RoleBindingResource is the name of the resource type that represents a role binding.
+  - RoleBindingSubjects is the names of the resource types that can be subjects in a role binding.
+  - RolebindingPermissionsPrefix generates the permissions sets to manage role bindings,
+  - GrantRelationship is the name of the relationship that connects a role binding to a resource.
```

Comment:
A few points here:
* I'm not sure `RoleResource`, `RoleBindingResource`, `RoleBindingPermissionsPrefix`, or `GrantRelationship` need to be configurable. What's gained by doing that?
* `RoleRelationshipSubject` doesn't feel like a descriptive name for what we're talking about here. What is a role relationship?
* What does it mean to own a role?

Reply:
- I do think `RoleResource` and `RoleBindingResource` should be configurable to avoid conflicted resource names, the rest I agree there's no need for the configurabilities, so I have them removed.
- ~`RoleRelationshipSubject`~ `RoleSubjectTypes`
- RoleOwners is the list of resource types that can own a role.
These resources should be (but not limited to) organizational resources
like tenant, organization, project, group, etc
When a role is owned by an entity, say a group, that means this role
will be available to perform role-bindings for resources that are owned
by this group and its subgroups.
The RoleOwners relationship is particularly useful to limit access to
custom roles. (this description is also added to the comment)